lib/events/pgevents/pgevents.go

```diff
@@ -0,0 +1,593 @@
+				retentionPeriod,
+			)
+			if err != nil {
+				return 0, trace.Wrap(err)
+			}
+
+			return tag.RowsAffected(), nil
+		})
+		if err != nil {
+			l.log.WithError(err).Error("Failed to execute periodic cleanup.")
+		} else {
+			l.log.WithField("deleted_rows", deleted).Debug("Executed periodic cleanup.")
+		}
+	}
+}
+
+var _ events.AuditLogger = (*Log)(nil)
+
+// EmitAuditEvent implements [events.AuditLogger].
+func (l *Log) EmitAuditEvent(ctx context.Context, event apievents.AuditEvent) error {
+	var sessionID uuid.UUID
+	if s := events.GetSessionID(event); s != "" {
+		u, err := uuid.Parse(s)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		sessionID = u
+	}
+
+	eventJSON, err := utils.FastMarshal(event)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	eventID := uuid.New()
+
+	// if an event with the same event_id exists, it means that we inserted it
+	// and then failed to receive the success reply from the commit
+	if _, err := pgcommon.RetryIdempotent(ctx, l.log, func() (struct{}, error) {
+		_, err := l.pool.Exec(ctx,
+			"INSERT INTO events (event_time, event_id, event_type, session_id, event_data)"+
+				" VALUES ($1, $2, $3, $4, $5)"+
+				" ON CONFLICT DO NOTHING",
+			event.GetTime().UTC(), eventID, event.GetType(), sessionID, eventJSON,
+		)
+		return struct{}{}, trace.Wrap(err)
+	}); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// searchEvents returns events within the time range, filtering (optionally) by
+// event types, session id, and a generic condition, limiting results by a count
+// and by a maximum size of the underlying json data of
+// events.MaxEventBytesInResponse, sorting by time, session id and event index
+// either ascending or descending, and returning an opaque, URL-safe string that
+// can be passed to the same function to continue fetching data.
+func (l *Log) searchEvents(
+	ctx context.Context,
+	fromTime, toTime time.Time,
+	eventTypes []string, cond *types.WhereExpr, sessionID string,
+	limit int, order types.EventOrder, startKey string,
+) ([]apievents.AuditEvent, string, error) {
+	var sessionUUID uuid.UUID
+	if sessionID != "" {
+		var err error
+		sessionUUID, err = uuid.Parse(sessionID)
+		if err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+	}
+
+	if limit <= 0 {
+		limit = defaults.EventsIterationLimit
+	}
+	if limit > defaults.EventsMaxIterationLimit {
+		return nil, "", trace.BadParameter("limit %v exceeds %v", limit, defaults.MaxIterationLimit)
+	}
+
+	var startTime time.Time
+	var startID uuid.UUID
+	if startKey != "" {
+		var err error
+		startTime, startID, err = fromStartKey(startKey)
+		if err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+	}
+
+	var condFn utils.FieldsCondition
+	if cond != nil {
+		var err error
+		condFn, err = utils.ToFieldsCondition(cond)
+		if err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+	}
+
+	var qb strings.Builder
+	qb.WriteString("DECLARE cur CURSOR FOR SELECT" +
```

Comment:
Doing string operation instead of using a parametrized question won't trash the internal DB cache?

Reply:
It's not that bad now that there's only 16 possibilities (a lot of which, I bet, are actually never exercised), but I'll leave a comment to run the query in `pgx.QueryExecModeExec` if we eventually decide to do server-side filtering with arbitrary `WhereCond`s.

Reply:
3d7521ac7889c8e457c96bb18ab7e948f270c62c